### PR TITLE
BUG: passing DataFrame to make_block silently raises

### PR DESCRIPTION
**Pull request by jbrockmendel**
ATM in the `except NotImplementedError:` branch of `_cython_agg_blocks`, we take a difference approach to the operation.  if that approach succeeds, we end up passing a DataFrame to `make_block` on L189, which will raise `ValueError`.  As a result, we'll end up falling back to python-space for the entire operation, which presumably entails a performance hit.

This fixes the incorrect passing of DataFrame, but the fix is kind of kludgy.  Suggestions welcome on how to improve it.

### CODE REVIEW

**Comment by WillAyd** [comment]
Hmm so you are saying this improves mixed block performance where some can be cythonized right?

This adds a little bit of logic so would be good to see that measured somehow to understand if it’s worth it

**Comment by jbrockmendel** [comment]
> Hmm so you are saying this improves mixed block performance where some can be cythonized right?

I'm having a hard time cooking up an example, but roughly what I have in mind is a case with 

1) a bunch of non-EA blocks and then 
2) one EA block such that 
  a) NotImplementedError is raised at L164 (L160 in PR) and 
  b) the fallback on L179 (L180 in PR) goes through OK, so that
  c) in master, we get a ValueError when passing `result` to `block.make_block` on L180

In such a case we would end up falling back to the non-cython implementations for _all_ blocks instead of just the one relevant EA block.

File: pandas/core/groupby/generic.py [hunk 1]
```diff
@@ -174,18 +175,36 @@ def _cython_agg_blocks(self, how, alt=None, numeric_only=True, min_count=-1):
...
+                s = obj.groupby(self.grouper)
                 try:
                     result = s.aggregate(lambda x: alt(x, axis=self.axis))
                 except TypeError:
                     # we may have an exception in trying to aggregate
                     # continue and exclude the block
                     deleted_items.append(locs)
                     continue
+
+                if is_object_dtype(block.dtype) and how in ["prod", "cumprod", "sum"]:
```
Comment (jreback) [inline pandas/core/groupby/generic.py]: a way around is to have .aggregate different exceptions for flow control (make some private ones that inherit from NotImplementedError); we do this in the cython aggregator now (well we raise NotImplementedError)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/groupby/generic.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/groupby/generic.py]: yah, distinguishing between our NotImplementedError vs numpy's is basically what #28198 does (albeit that is only in one place).  I think the groupby catching/raising is complicated enough that we need to do a few passes clearing up what gets caught where, then go through to do the catching in fewer places

**Comment by jbrockmendel** [comment]
Closing to clear the queue.  Will revisit.

**Pull request closed without merging**